docs/source/cpp/acero/substrait.rst

```diff
@@ -0,0 +1,248 @@
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. default-domain:: cpp
+.. highlight:: cpp
+.. cpp:namespace:: arrow::engine::substrait
+
+.. _acero-substrait:
+
+==========================
+Using Acero with Substrait
+==========================
+
+In order to use Acero you will need to create an execution plan.  This is the
+model that describes the computation you want to apply to your data.  Acero has
+its own internal representation for execution plans but most users should not
+interact with this directly as it will couple their code to Acero.
+
+`Substrait <https://substrait.io>`_ is an open standard for execution plans.
+Acero implements the Substrait "consumer" interface.  This means that Acero can
+accept a Substrait plan and fulfill the plan, loading the requested data and
+applying the desired computation.  By using Substrait plans users can easily
+switch out to a different execution engine at a later time.
+
+Substrait Conformance
+---------------------
+
+Substrait defines a broad set of operators and functions for many different
+situations and it is unlikely that Acero will ever completely satisfy all
+defined Substrait operators and functions.  To help understand what features
+are available the following sections define which features have been currently
+implemented in Acero and any caveats that apply.
+
+Plans
+^^^^^
+
+ * A plan should have a single top-level relation.
+ * The consumer is currently based on version 0.20.0 of Substrait.
+   Any features added that are newer will not be supported.
+ * Due to a breaking change in 0.20.0 any Substrait plan older than 0.20.0
+   will be rejected.
+
+Extensions
+^^^^^^^^^^
+
+ * If a plan contains any extension type variations it will be rejected.
+ * Advanced extensions can be provided by supplying a custom implementation of
+   :class:`arrow::engine::ExtensionProvider`.
+
+Relations (in general)
+^^^^^^^^^^^^^^^^^^^^^^
+
+ * Any relation not explicitly listed below will not be supported
+   and will cause the plan to be rejected.
+
+Read Relations
+^^^^^^^^^^^^^^
+
+ * The ``projection`` property is not supported and plans containing this
+   property will be rejected.
+ * The ``VirtualTable`` and ``ExtensionTable`` read types are not supported.
+   Plans containing these types will be rejected.
+ * Only the parquet and arrow file formats are currently supported.
+ * All URIs must use the ``file`` scheme
+ * ``partition_index``, ``start``, and ``length`` are not supported.  Plans containing
+   non-default values for these properties will be rejected.
+ * The Substrait spec requires that a ``filter`` be completely satisfied by a read
+   relation.  However, Acero only uses a read filter for pushdown projection and
+   it may not be fully satisfied.  Users should generally attach an additional
+   filter relation with the same filter expression after the read relation.
```

Comment:
Side question, but wondering while reading this: if that's what the spec of substrait requires, shouldn't the translation of a substrait plan to Acero ExecPlan inject such an additional filter node to ensure we actually fulfill that requirement? (instead of requiring the user to do that explicitly in their Substrait plan)

Reply:
I think that's a fair point.  Let's address in a follow-up.